// angular-l10n/src/models/utils/base-directive.js
import { BFS } from './bfs';
/**
 * @abstract
 */
export var BaseDirective = (function () {
    /**
     * @param {?} el
     * @param {?} renderer
     */
    function BaseDirective(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.breadthFirstSearch = new BFS();
    }
    /**
     * @return {?}
     */
    BaseDirective.prototype.ngAfterViewInit = function () {
        this.element = this.el.nativeElement;
        this.renderNode = this.breadthFirstSearch.getTargetNode(this.element);
        this.nodeValue = this.renderNode != null ? (this.renderNode.nodeValue) : null;
        if (this.element.hasChildNodes()) {
            this.key = this.getText();
        }
        else if (this.element.hasAttribute("value")) {
            this.key = this.element.getAttribute("value");
        }
        if (!!this.key) {
            this.setup();
        }
    };
    ;
    /**
     * @abstract
     * @return {?}
     */
    BaseDirective.prototype.setup = function () { };
    /**
     * @abstract
     * @return {?}
     */
    BaseDirective.prototype.replace = function () { };
    /**
     * @param {?} value
     * @return {?}
     */
    BaseDirective.prototype.setText = function (value) {
        if (!!this.nodeValue) {
            this.renderer.setText(this.renderNode, this.nodeValue.replace(this.key, value));
        }
        else if (this.element.hasAttribute("value")) {
            this.renderer.setElementAttribute(this.element, "value", value);
        }
    };
    /**
     * @return {?}
     */
    BaseDirective.prototype.getText = function () {
        return this.nodeValue != null ? this.nodeValue.trim() : null;
    };
    return BaseDirective;
}());
function BaseDirective_tsickle_Closure_declarations() {
    /** @type {?} */
    BaseDirective.prototype.key;
    /** @type {?} */
    BaseDirective.prototype.element;
    /** @type {?} */
    BaseDirective.prototype.renderNode;
    /** @type {?} */
    BaseDirective.prototype.nodeValue;
    /** @type {?} */
    BaseDirective.prototype.breadthFirstSearch;
    /** @type {?} */
    BaseDirective.prototype.el;
    /** @type {?} */
    BaseDirective.prototype.renderer;
}
