// angular-l10n/src/models/localization/browser.js
/**
 * Manages the browser for storage, navigator and Intl.
 */
export var Browser = (function () {
    /**
     * @param {?} locale
     */
    function Browser(locale) {
        this.locale = locale;
        this.hasCookie = typeof navigator.cookieEnabled !== "undefined" && navigator.cookieEnabled;
        this.hasLocalStorage = typeof Storage !== "undefined";
    }
    /**
     * @param {?} name
     * @return {?}
     */
    Browser.prototype.readStorage = function (name) {
        var /** @type {?} */ value;
        if (!this.storageIsDisabled) {
            if (this.locale.configuration.localStorage && this.hasLocalStorage) {
                value = this.getLocalStorage(name);
            }
            else if (this.hasCookie) {
                value = this.getCookie(name);
            }
        }
        return value;
    };
    /**
     * @param {?} name
     * @param {?} value
     * @return {?}
     */
    Browser.prototype.writeStorage = function (name, value) {
        if (!this.storageIsDisabled) {
            if (this.locale.configuration.localStorage && this.hasLocalStorage) {
                this.setLocalStorage(name, value);
            }
            else if (this.hasCookie) {
                this.setCookie(name, value);
            }
        }
    };
    /**
     * @return {?}
     */
    Browser.prototype.getBrowserLanguage = function () {
        var /** @type {?} */ browserLanguage;
        if (typeof navigator.language !== "undefined") {
            browserLanguage = navigator.language;
        }
        if (browserLanguage != null) {
            var /** @type {?} */ index = browserLanguage.indexOf("-");
            if (index != -1) {
                browserLanguage = browserLanguage.substring(0, index);
            }
        }
        return browserLanguage;
    };
    /**
     * @param {?} name
     * @return {?}
     */
    Browser.prototype.getLocalStorage = function (name) {
        return localStorage.getItem(name);
    };
    /**
     * @param {?} name
     * @return {?}
     */
    Browser.prototype.getCookie = function (name) {
        var /** @type {?} */ result = new RegExp("(?:^|; )" +
            encodeURIComponent(name) +
            "=([^;]*)").exec(document.cookie);
        return result ? result[1] : null;
    };
    /**
     * @param {?} name
     * @param {?} value
     * @return {?}
     */
    Browser.prototype.setLocalStorage = function (name, value) {
        localStorage.setItem(name, value);
    };
    /**
     * @param {?} name
     * @param {?} value
     * @return {?}
     */
    Browser.prototype.setCookie = function (name, value) {
        var /** @type {?} */ expires = "";
        if (this.locale.configuration.cookiesExpirationDays != null) {
            var /** @type {?} */ expirationDate = new Date();
            expirationDate.setTime(expirationDate.getTime() +
                (this.locale.configuration.cookiesExpirationDays * 24 * 60 * 60 * 1000));
            expires = "; expires=" + expirationDate.toUTCString();
        }
        document.cookie = name + "=" + value + expires + "; path=/";
    };
    return Browser;
}());
function Browser_tsickle_Closure_declarations() {
    /** @type {?} */
    Browser.prototype.storageIsDisabled;
    /** @type {?} */
    Browser.prototype.hasCookie;
    /** @type {?} */
    Browser.prototype.hasLocalStorage;
    /** @type {?} */
    Browser.prototype.locale;
}
