// angular-l10n/src/modules/localization.module.js
import { NgModule } from '@angular/core';
import { TranslationModule } from './translation.module';
import { LocaleService } from '../services/locale.service';
import { TranslationService } from '../services/translation.service';
import { LocaleDatePipe } from '../pipes/locale-date.pipe';
import { LocaleDecimalPipe, LocalePercentPipe, LocaleCurrencyPipe } from '../pipes/locale-number.pipe';
import { LocaleDateDirective } from '../directives/locale-date.directive';
import { LocaleDecimalDirective, LocalePercentDirective, LocaleCurrencyDirective } from '../directives/locale-number.directive';
export var LocalizationModule = (function () {
    function LocalizationModule() {
    }
    /**
     * Use in AppModule: new instances of LocaleService & TranslationService.
     * @return {?}
     */
    LocalizationModule.forRoot = function () {
        return {
            ngModule: LocalizationModule,
            providers: [LocaleService, TranslationService]
        };
    };
    /**
     * Use in features modules with lazy loading: new instance of TranslationService.
     * @return {?}
     */
    LocalizationModule.forChild = function () {
        return {
            ngModule: LocalizationModule,
            providers: [TranslationService]
        };
    };
    LocalizationModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        LocaleDatePipe,
                        LocaleDecimalPipe,
                        LocalePercentPipe,
                        LocaleCurrencyPipe,
                        LocaleDateDirective,
                        LocaleDecimalDirective,
                        LocalePercentDirective,
                        LocaleCurrencyDirective
                    ],
                    imports: [
                        TranslationModule
                    ],
                    exports: [
                        TranslationModule,
                        LocaleDatePipe,
                        LocaleDecimalPipe,
                        LocalePercentPipe,
                        LocaleCurrencyPipe,
                        LocaleDateDirective,
                        LocaleDecimalDirective,
                        LocalePercentDirective,
                        LocaleCurrencyDirective
                    ]
                },] },
    ];
    /** @nocollapse */
    LocalizationModule.ctorParameters = function () { return []; };
    return LocalizationModule;
}());
function LocalizationModule_tsickle_Closure_declarations() {
    /** @type {?} */
    LocalizationModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    LocalizationModule.ctorParameters;
}
