// angular-l10n/src/modules/locale-validation.module.js
import { NgModule } from '@angular/core';
import { LocaleValidation } from '../services/locale-validation';
import { LocaleNumberValidatorDirective } from '../directives/locale-number-validator.directive';
export var LocaleValidationModule = (function () {
    function LocaleValidationModule() {
    }
    /**
     * Use in AppModule: new instance of LocaleValidation.
     * @return {?}
     */
    LocaleValidationModule.forRoot = function () {
        return {
            ngModule: LocaleValidationModule,
            providers: [LocaleValidation]
        };
    };
    LocaleValidationModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        LocaleNumberValidatorDirective
                    ],
                    exports: [
                        LocaleNumberValidatorDirective
                    ]
                },] },
    ];
    /** @nocollapse */
    LocaleValidationModule.ctorParameters = function () { return []; };
    return LocaleValidationModule;
}());
function LocaleValidationModule_tsickle_Closure_declarations() {
    /** @type {?} */
    LocaleValidationModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    LocaleValidationModule.ctorParameters;
}
